src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc

```diff
@@ -365,6 +365,86 @@ FileWatcherCertificateProvider::ReadIdentityKeyCertPairFromFiles(
   return absl::nullopt;
 }
 
+DataWatcherCertificateProvider::DataWatcherCertificateProvider(
+    std::string root_certificate,
+    grpc_core::PemKeyCertPairList pem_key_cert_pairs)
+    : StaticDataCertificateProvider(std::move(root_certificate), std::move(pem_key_cert_pairs)) {}
+
+absl::Status DataWatcherCertificateProvider::ReloadRootCertificate(
+    const std::string& root_certificate) {
+  if (root_certificate.empty()) {
+    return absl::InvalidArgumentError("Root Certificate string is empty.");
+  }
+  grpc_core::MutexLock lock(&mu_);
+  if (root_certificate_ == root_certificate) {
+    return absl::InvalidArgumentError("Root Certificate has not changed.");
+  }
+  root_certificate_ = root_certificate;
+  ExecCtx exec_ctx;
+  grpc_error_handle root_cert_error = GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+      "Unable to get latest root certificates.");
+  for (const auto& p : watcher_info_) {
+    const std::string& cert_name = p.first;
+    const WatcherInfo& info = p.second;
+    absl::optional<std::string> root_to_report;
+    if (info.root_being_watched) {
+      root_to_report = root_certificate_;
+    }
+    if (root_to_report.has_value()) {
+      distributor_->SetKeyMaterials(cert_name, std::move(root_to_report),
+                                    absl::nullopt);
+    }
+    if (info.root_being_watched && root_certificate_.empty()) {
+      distributor_->SetErrorForCert(cert_name, GRPC_ERROR_REF(root_cert_error),
+                                    absl::nullopt);
+    }
+  }
+  GRPC_ERROR_UNREF(root_cert_error);
+  return absl::OkStatus();
+}
+
+absl::Status DataWatcherCertificateProvider::ReloadKeyCertificatePair(
+    grpc_core::PemKeyCertPairList pem_key_cert_pairs) {
+  if (pem_key_cert_pairs.empty()) {
+    return absl::InvalidArgumentError("Empty Key-Cert pair list.");
+  }
+  if (pem_key_cert_pairs == pem_key_cert_pairs_) {
+    return absl::InvalidArgumentError(
+        "The Key-Cert pair list has not changed.");
+  }
+  for (unsigned long i = 0; i < pem_key_cert_pairs.size(); i++) {
```

Comment:
Can we reuse GetValidKeyCertPairList function here instead? So we won't need to repeat code here.
This means we will have to update the function signature of GetValidKeyCertPairList, make it return a status and accept pair list as a reference. If the function returns an OK status the input was valid, and we just use that. If we get a non-OK status we return that status.
In StaticDataCertificateProvider, if GetValidKeyCertPairList returns not-OK status, we initialize to an empty list. For an OK status, we move pem_key_cert_pairs to member variable.
Wdyt?

Reply:
This sounds great!